add PR template (#27392)
hashicorp/vault: .github/pull_request_template.md

### Description
What does this PR do?

### TODO

> - [ ] Make sure there's a changelog. Brand new features have a differently

Reviewer: This template will apply globally. I think (for the most part) we may want to have different guidance for, say, a docs PR. I _think_ we can address this in the future by adding a `.github/docs/pull_request_template.md` and so on -- the sky's the limit!

Author: I can add another template for docs PRs

  formatted changelog than other changelogs, so pay attention to that. If this
  PR is the CE portion of an ENT PR, put the changelog here, _not_ in your ENT
  PR.
- [ ] Add a milestone. If this change is going to be backported, each backport
  PR should receive the milestone for the exact version it's being backported
  to. For this PR, just pick the biggest version.

> - [ ] If this PR is the CE portion of an ENT change, and that ENT change is

Reviewer: It might be worthwhile to try to remind people to add:

1. A reference to the associated Jira, if there is one: (i.e.`Resolves: VAULT-12358132135`)
2. A link to the associated RFC (if there is one).
3. A link to the associated (ideally, approved) Enterprise PR.

Reviewer: Oh and maybe some guidance about Squash Merge wouldn't be too bad.

Author: I added your first 3 bullet points. It was brought up that CE PRs can only be squash merged, so the last point isn't necessary.

  getting backported to N-2, use the new style `backport/ent/x.x.x+ent` labels
  instead of the old style `backport/x.x.x` labels.
- [ ] If this PR is a CE only change, it can only be backported to N, so use
  the normal `backport/x.x.x` label (there should be only 1).

> - [ ] If this PR either 1) removes a public function OR 2) changes the signature

Reviewer: Do you think it would be helpful to link to the Confluence page? https://hashicorp.atlassian.net/wiki/spaces/VAULT/pages/2295365741/The+Vault+Development+Cycle#Enterprise-or-OSS

This raises another question: do we want this to apply to community-submitted PRs? The only way I can think of making a distinction would be to instead push the logic into CI, but that doesn't seem great.

Author: This is a HC employee only thing. Community members can't check the ENT repo. The Confluence page doesn't seem like it adds much clarity to me. In my mind, this should be a self-explanatory checklist that people can just eyeball and check off. If we're requiring background reading to complete the checklist, it's a bad checklist and people probably won't follow it.

That said, if this item doesn't feel self-explanatory to you, or should be a in a greater level of detail, I'm certainly open to re-wording it.

### Description
What does this PR do?

### TODO only if you're a HashiCorp employee
- [ ] **Changelog:** Make sure there is one, or the `pr/no-changelog` label is
  applied. Brand new features have a differently
  formatted changelog than other changelogs, so pay attention to that. If this
  PR is the CE portion of an ENT PR, put the changelog here, _not_ in your ENT
  PR. Feel free to refer to the [instructions](https://github.com/hashicorp/vault/blob/main/CONTRIBUTING.md#changelog-entries) on changelog formatting if necessary.

> - [ ] **Labels:** If this PR is the CE portion of an ENT change, and that ENT change is
>   getting backported to N-2, use the new style `backport/ent/x.x.x+ent` labels
>   instead of the old style `backport/x.x.x` labels.

Reviewer 2: ```suggestion
- [ ] **Labels:** If this PR is part of a change is
  getting backported to N-2, use the `backport/ent/x.x.x+ent` style labels
  instead of the `backport/x.x.x` style labels.
```

- [ ] **Labels:** If this PR is a CE only change, it can only be backported to N, so use
  the normal `backport/x.x.x` label (there should be only 1).
- [ ] **ENT Breakage:** If this PR either 1) removes a public function OR 2) changes the signature
  of a public function, even if that change is in a CE file, _double check_ that
  applying the patch for this PR to the ENT repo and running tests doesn't
  break any tests. Sometimes ENT only tests rely on public functions in CE
  files.

> - [ ] **Jira:** If this change has an associated Jira, please link it in the description.

Reviewer 2: ```suggestion
- [ ] **Jira:** If this change has an associated Jira, it's linked it in the description, commit message, or branch name.
```

Reviewer 2: Woops, typo. Resubmitting.

### Description
What does this PR do?

### TODO if you're a community member

> - [ ] Make sure there's a changelog. This is a description of what changed from

Reviewer 3: Do we often have to ask community members to fix their changelog formatting? Perhaps linking to [this part](https://github.com/hashicorp/vault/blob/main/CONTRIBUTING.md#changelog-entries) of CONTRIBUTING.md would be helpful here since it shows exactly what a changelog should look like.

Reviewer 2: It's not uncommon. A lot of community PRs do include a changelog, but sometimes on old PRs I sometimes just add the changelog for them (since asking them to come back after months or years to add it can feel a little wrong).

Still, I think that linking to the contributing.md would be valuable, since this description doesn't say what a changelog is. I do wonder if this kind of information should go in the failing changelog GHA instead, though. I think it's best (where possible) to not duplicate

Here's an example output from a failing changelog entry today:
```
looking for changelog file changelog/25080.txt
Not found.
looking for changelog file matching changelog/_go-ver-*.txt
Not found.

Did not find a changelog entry named changelog/25080.txt
If your changelog file is correct, skip this check with the 'pr/no-changelog' label
Reference - https://github.com/hashicorp/vault/pull/10363 and https://github.com/hashicorp/vault/pull/11894
```

Reviewer 4: In addition, we often have to remind community members to add the changelog file. I think we should either link to the contributing.md for changelog specifics, because many folks don't know how we want that done, how it should be formatted, and the like. The wording here might also be too vague in general, and I'll noodle a bit on how we might make it clearer for the community.

Author: I've updated this item to include a link to those changelog instructions.

  a user's point of view.

### TODO if you're a HashiCorp employee

> - [ ] Make sure there's a changelog. Brand new features have a differently

Reviewer 2: I think the changelog check should obviate the need for this checkbox. It's worth noting that not all changes require changelogs, either.

I also don't know that this is always true?
```
If this
  PR is the CE portion of an ENT PR, put the changelog here, _not_ in your ENT
  PR.
```

If it's for an enterprise feature, my understanding is that the changelog goes in the ent PR so it signals it's for ent only. For example, if an update to lease count quotas has a CE change, the changelog should be on the ent PR.

Reviewer 5: The official guidance doesn't speak to changes that are purely ent-only, but is:

> If you are making a change for an issue and it requires an enterprise PR and a CE PR, it’s preferable to make the changelog entry in the CE repository with the changelog entry name corresponding to the CE PR number, not the enterprise PR number. Having the changelog entry in the CE repo lets us link to the CE PR, whereas in the enterprise repo we cannot link to the PR.

Reviewer 2: Oh, huh, TIL. I guess I've always been doing it counter to that advice for features that are primarily in ent

Author: I like including this here because although yes, there is a GHA check for a changelog, the GHA check does not remind you that features need to have a changelog that is formatted differently than other changelogs, something that myself and others have gotten dinged on in the past.

Reviewer 2: Could we move that information to the GHA instead? It should just be updating this bit, which I think is a little outdated compared to the `contributing.md` link:

```
            if [ -z "$toolchain_files" ]; then
              echo "Not found."
              echo ""
              echo "Did not find a changelog entry named ${expected_changelog_file}"
              echo "If your changelog file is correct, skip this check with the 'pr/no-changelog' label"
              echo "Reference - https://github.com/hashicorp/vault/pull/10363 and https://github.com/hashicorp/vault/pull/11894"
              exit 1
            fi
```

Author: That will only show up in GHA if there's a problem in the changelog though, no? If I include a changelog with my PR, the GHA check is satisfied and nothing is logged, even if my changelog is incorrectly formatted.

Reviewer 2: That would show up if there isn't a changelog, yeah. You're right, that wouldn't show up if I'd incorrectly formatted a feature, but we do have a lot of checks for incorrectly formatted changelogs there today -- it'd fit neatly. My bash-fu isn't great but it'd probably be something like this:

```
          elif grep -q ':feature$' "$changelog_files"; then
            # Pseudocode: if second line doesn't start with **
            echo "Incorrect feature formatting, make sure you do X instead"
            exit 1
```

If you'd like, I can play around with this a little and see if I can get something working. I've also done this incorrectly in the past and it'd be great to add this to the list of things we check. If you're not feeling up to doing this now then I don't think it should block the PR template, but it'd be great if we could automate that part of the checklist away

Author: It's not something I'm going to focus on in this PR, since this is really just about the template, but if you want to follow this with a PR of your own that automates a formatting check and removes an item from the checklist, that sounds great to me.

To be clear, I'm not a checklist enthusiast. In my happy world, we wouldn't have checklists at all and automation would handle literally everything. But I think this is better than nothing at all until we get to that point.

Maybe if enough people are irritated with the checklist, it will spur the creation of additional automation 😅

Reviewer 2: Totally agreed! This was useful to understand the root of the desire for the checklist here, and knowing that you'd be happy if we automated it is great news to me. Checklists are easy, effective and give us a great list of things that we could automate to make our lives even easier

  formatted changelog than other changelogs, so pay attention to that. If this
  PR is the CE portion of an ENT PR, put the changelog here, _not_ in your ENT
  PR.

> - [ ] Add a milestone. If this change is going to be backported, each backport

Author: I removed this

  PR should receive the milestone for the exact version it's being backported
  to. For this PR, just pick the biggest version.

> - [ ] If this PR is the CE portion of an ENT change, and that ENT change is

Reviewer 3: I was talking recently with Marc about his time as a pilot, where checklists are used extensively. 

He explained how the pilot community has identified that checklists work best when they're a clear action, and preferably just a short keyword or two you can read at a glance, otherwise human nature is to feel cognitively overwhelmed by all the text and gloss over them.

However, I also agree with your assessment that we want the important details right here in the PR template, so I wonder if we can combine both of those philosophies by making sure that each of your sentences here starts with a bold obvious keyword.

Like:

[ ] **Labels**: If this PR is the CE portion of an ENT change, and that ENT change is ... etc.

Author: Sure, will do

  getting backported to N-2, use the new style `backport/ent/x.x.x+ent` labels
  instead of the old style `backport/x.x.x` labels.
- [ ] If this PR is a CE only change, it can only be backported to N, so use
  the normal `backport/x.x.x` label (there should be only 1).
- [ ] If this PR either 1) removes a public function OR 2) changes the signature
  of a public function, even if that change is in a CE file, _double check_ that
  applying the patch for this PR to the ENT repo and running tests doesn't
  break any tests. Sometimes ENT only tests rely on public functions in CE

>   files.

Reviewer 3: How does that stubmaker thing fit in here? Should there be a checkbox for that?

Author: Stubmaker is not widely adopted so I'm not sure if it makes sense to add here or not. If so, someone else should add that, as I've never used it.

> - [ ] If this change has an associated Jira, please link it in the description.

Author: Both approaches seem fine

> - [ ] If this change has an associated RFC, please link it in the description.

Author: It's not a great state to be in, but I've been in it before, and I doubt I'm the only one. I feel like there's a tension here between including the bare minimum of information in this checklist, and hoping it's enough to prove useful, and including the maximum amount of information in this checklist, and hoping people bother to fill it out. Anything that makes my job as a reviewer easier is good to have imo and I struggle to think of a time when I thought `ugh why is this person giving me all of this background information, do not want`.

> - [ ] If this change has an associated ENT PR, please link it in the

Reviewer 2: For a PR that touches ent, usually I find that the basis PR (and 'main' PR) is enterprise and the CE PR is linked to from that, as opposed to the other way around. This also could be something we add into the `no-oss-changes` GHA instead?

Author: I don't think it hurts to have links in both directions. I've frequently been asked to review a CE PR that had an associated ENT PR and there was no info on what the ENT PR was, which necessitated a lot of digging through ENT PRs.

  description.
- [ ] If this is a normal PR, please remember to merge your PRs using the

>   `Squash and merge` button and _NOT_ the `Merge` button that creates a separate

Reviewer 3: In the interest of keeping checklist items short and clear, I think you can get rid of that last explanatory sentence since it sounds like no one needs to ever hit that button.

Reviewer 2: I don't know that we need this check at all, actually, since on CE, the only option is squash and merge

Author: I've removed this
